lib/ask/survey_respondents_canceller.ex

```diff
@@ -1,111 +1,65 @@
-defmodule Ask.RespondentsCancellerProducer do
+defmodule Ask.SurveyCanceller do
   use Ecto.Schema
+  use GenServer
   import Ecto.Query
-  alias Ask.{Respondent, Survey, Repo, ActivityLog, Project}
-  alias Ecto.Multi
-  use GenStage
 
-  def start_link(initial) do
-    GenStage.start_link(__MODULE__, initial, name: __MODULE__)
-  end
+  alias Ask.{
+    ActivityLog,
+    Logger,
+    Project,
+    Repo,
+    Respondent,
+    Survey
+  }
 
-  def init(survey_id) when not is_nil(survey_id) do
-    state = %{survey_ids: [survey_id], last_updated_respondent_id: 0}
-    {:producer, state}
-  end
+  alias Ask.Runtime.{
+    Session,
+    SurveyCancellerSupervisor
+  }
 
-  def init(survey_id) when is_nil(survey_id) do
-    query =
-      from(
-        s in Survey,
-        where: s.state == :cancelling,
-        select: s.id
-      )
-
-    survey_ids =
-      query
-      |> Repo.all()
-
-    case survey_ids do
-      [] ->
-        :ignore
-
-      _ ->
-        state = %{survey_ids: survey_ids, last_updated_respondent_id: 0}
-        {:producer, state}
-    end
-  end
+  alias Ecto.Multi
 
-  def handle_demand(_demand, state) do
-    respondent_ids = get_respondents_for_update(state)
-    handle_respondents(respondent_ids, state)
+  def start_link(survey_id) do
+    GenServer.start_link(__MODULE__, survey_id)
   end
 
-  def handle_respondents(respondent_ids, state) when length(respondent_ids) > 0 do
-    last_id = List.last(respondent_ids)
-    new_state = %{survey_ids: state.survey_ids, last_updated_respondent_id: last_id}
-    {:noreply, respondent_ids, new_state}
+  def cancel(survey_id) do
+    SurveyCancellerSupervisor.start_cancelling(survey_id)
```

Comment:
Good one. I left it in the Canceller because that's where it was, but it makes more sense to move it to the Supervisor 👍

The Canceller is now kind of "hidden" behind the supervisor - the "public" interface for the app to use is through the Supervisor.